### PR TITLE
CSHARP-5319: Remove AsDateTime and similar from BsonValue

### PR DESCRIPTION
**Pull request by papafe**

### CODE REVIEW

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 1]
```diff
@@ -227,15 +218,6 @@ public int AsInt32
             get { return ((BsonInt32)this).Value; }
         }
 
-        /// <summary>
-        /// Casts the BsonValue to a DateTime in the local timezone (throws an InvalidCastException if the cast is not valid).
-        /// </summary>
-        [Obsolete("Use ToLocalTime instead.")]
-        public DateTime AsLocalTime
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:234]: I think this was obsoleted by mistake and we should keep the property and remove the `[Obsolete]` attribute.

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 2]
```diff
@@ -341,15 +314,6 @@ public string AsString
             get { return ((BsonString)this).Value; }
         }
 
-        /// <summary>
-        /// Casts the BsonValue to a DateTime in UTC (throws an InvalidCastException if the cast is not valid).
-        /// </summary>
-        [Obsolete("Use ToUniversalTime instead.")]
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:347]: I think this was obsoleted by mistake and we should keep the property and remove the `[Obsolete]` attribute.

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 3]
```diff
@@ -252,15 +234,6 @@ public bool? AsNullableBoolean
             get { return (BsonType == BsonType.Null) ? null : (bool?)AsBoolean; }
         }
 
-        /// <summary>
-        /// Casts the BsonValue to a Nullable{DateTime} (throws an InvalidCastException if the cast is not valid).
-        /// </summary>
-        [Obsolete("Use ToNullableUniversalTime instead.")]
-        public DateTime? AsNullableDateTime
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:259]: This property should be replaced with two properties: `AsNullableLocalTime` and `AsNullableUniversalTime`.

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 4]
```diff
@@ -178,15 +178,6 @@ public byte[] AsByteArray
             get { return ((BsonBinaryData)this).Bytes; }
         }
 
-        /// <summary>
-        /// Casts the BsonValue to a DateTime in UTC (throws an InvalidCastException if the cast is not valid).
-        /// </summary>
-        [Obsolete("Use ToUniversalTime instead.")]
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:184]: This obsoletion was proper and the member can be deleted.

The error message was wrong though. It should have been "Use AsUniversalTime" instead.

**Review by rstam — CHANGES_REQUESTED** [review]
There is a discussion going on about what exactly we should do.

I am proposing that we should keep `AsLocalTime` and `AsUniversalTime` because that it is the existing pattern established by all our other `AsXyz` properties.

**Comment by rstam** [comment]
Given our time zone difference and that I'm out of the office tomorrow I have prototyped in code what we were discussing earlier today:

https://github.com/rstam/mongo-csharp-driver/commits/csharp5319/

Summary of proposed changes:

Remove `AsDateTime` (use `AsUniversalTime` instead)
Remove `AsNullableDateTime` (use `AsNullableUniversalTime` instead)

Keep `AsLocalTime` (remove `[Obsolete]` attribute)
Keep `AsUniversalTime` (remove `[Obsolete]` attribute)

Add `AsNullableLocalTime`
Add `AsNullableUniversalTime`

Leave any existing "to date" methods alone.

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 5]
```diff
@@ -245,20 +228,20 @@ public long AsInt64
         }
 
         /// <summary>
-        /// Casts the BsonValue to a Nullable{Boolean} (throws an InvalidCastException if the cast is not valid).
+        /// Casts the BsonValue to a DateTime in the local timezone (throws an InvalidCastException if the cast is not valid).
         /// </summary>
-        public bool? AsNullableBoolean
+        [EditorBrowsable(EditorBrowsableState.Never)]
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:233]: I think we want `DebuggerBrowsable` here.

Also, we want it for ALL `AsXyz` properties, not just the DateTime ones.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs (hunk 5 above)
Comment (papafe) [reply src/MongoDB.Bson/ObjectModel/BsonValue.cs:233]: Yes, you're right, `DebuggerBrowsable` is the correct one, I've misunderstood the situation

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 6]
```diff
@@ -77,6 +82,7 @@ public BsonBinaryData AsBsonBinaryData
         /// <summary>
         /// Casts the BsonValue to a BsonDateTime (throws an InvalidCastException if the cast is not valid).
         /// </summary>
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:85]: Missing attribute on line 94.

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs [hunk 7]
```diff
@@ -165,6 +180,7 @@ public BsonUndefined AsBsonUndefined
         /// <summary>
         /// Casts the BsonValue to a BsonValue (a way of upcasting subclasses of BsonValue to BsonValue at compile time).
         /// </summary>
+        [DebuggerBrowsable(DebuggerBrowsableState.Never)]
```
Comment (rstam) [inline src/MongoDB.Bson/ObjectModel/BsonValue.cs:183]: Missing attribute on line 192.

**Review by rstam — CHANGES_REQUESTED** [review]

File: src/MongoDB.Bson/ObjectModel/BsonValue.cs (hunk 7 above)
Comment (papafe) [reply src/MongoDB.Bson/ObjectModel/BsonValue.cs:184]: 🤦 just added

**Review by rstam — APPROVED** [review]
LGTM

**Pull request merged**